README.md
rosmini - A simplified version of ROS for education purposes


CHANGELOG.md
Change Log
===========

0.0.1 - Initial release

rosmini/rosminicore.py

from collections import defaultdict
import threading
import socket

    
    

class rosminicore_organizer():
    topic_dict = defaultdict(bool) # "/topic" -> (ip, port)
    ports = []
    
    def add_topic(self, topic, ip):
        if(topic in self.topic_dict):
            return self.topic_dict[topic]
        port = self.ports[-1] + 1 if len(self.ports) > 0 else 9000
        if(port > 9999):
            port = 9000
        self.topic_dict[topic] = (ip, port)
        self.ports.append(port)
        return self.topic_dict[topic]

        
class message_class:
    len_start_bytes = 8
    start_bytes = b'rosmini!'
    len_header_bytes = 4
    # header bytes should be the len of the topic name
    start_bytes_read_topics = 'rmtopic!'




class rosminicore_receive():
    def __init__(self, ip="", port=6464):
        UDP_IP = ip
        UDP_PORT = port

        self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM) 
        self.sock.bind((UDP_IP, UDP_PORT))
    
    def read(self):
        data, addr = self.sock.recvfrom(message_class.len_start_bytes, socket.MSG_WAITALL)
        print(str(addr) + ' is trying to connect')
        if data:
            if data.decode() == message_class.start_bytes.decode():
                data, addr = self.sock.recvfrom(message_class.len_header_bytes, socket.MSG_WAITALL)
                if data:
                    len_topic = int(data.decode())
                    print('Topic length: ' + str(len_topic))
                    print(data)
                    data, addr = self.sock.recvfrom(len_topic, socket.MSG_WAITALL)
                    if data:
                        topic = str(data.decode())
                        print(str(addr) + ' registered on topic ' + topic)
                        ip, port = rosminicore_topics.add_topic(topic, addr[0])
                        message = str((ip, port))
                        len_message = len(message)
                        message_bytes = bytes(message)
                        self.sock.sendto(bytes(str(len_message)).ljust(4, b' '), addr)
                        self.sock.sendto(message_bytes, addr)
            elif data.decode() == message_class.start_bytes_read_topics.decode():
                print(rosminicore_organizer.topic_dict.items())
                message = str(rosminicore_topics.topic_dict.items())
                len_message = len(message)
                self.sock.sendto(bytes(len_message), addr)
                self.sock.sendto(bytes(message), addr)
                    
    
    def close(self):
        self.sock.close()



lock = threading.Lock()
rosminicore_topics = rosminicore_organizer()
import time

def main():
    print(socket.gethostname())

    rosminicore_connection = rosminicore_receive(socket.gethostname(), 6464)
            
    def listen_trigger_data():
        global trigger
        while True:
            ok = rosminicore_connection.read()



    s = threading.Thread(target=listen_trigger_data)
    s.daemon = True
    s.start()

    print('Starting')
    while True:
        try:
            time.sleep(1)
        except KeyboardInterrupt:
            print('Closing')
            rosminicore_connection.close()
            exit()
        except:
            print('Closing')
            rosminicore_connection.close()
            exit()

        
        


if __name__ == "__main__":
    main()

rosmini/rostopic_list.py
import socket
import ast


ANY = "0.0.0.0"
SENDERPORT = 32000
MCAST_ADDR = "237.252.249.227"


def communicate_rosminicore(sock):
    sock.sendall(b'rmtopic!')
    data, rosminicore_addr = sock.recvfrom(4, socket.MSG_WAITALL)
    if data:
        len_topic = int(data.decode())
        data, rosminicore_addr = sock.recvfrom(len_topic, socket.MSG_WAITALL)
        if data:
            topic = ast.literal_eval(data.decode())
            print(str(topic))



class TopicList:

    def __init__(self):

        sock_rosminicore = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        # sock_rosminicore.bind((socket.gethostname(), 6464))
        sock_rosminicore.connect((socket.gethostname(), 6464))
        communicate_rosminicore(sock_rosminicore)

if __name__ == '__main__':
    pub = TopicList()

rosmini/rosminicore_tcp.py
# implement a roscore like node
# it should be able to start a socket server and receive data from N clients
# it receives the name of a topic and if it is subbing or publishing
# if it is subbing, sends back the IP and port of the node that publishes the topic
# if it is publishing, it should remember the IP and port of the node

import socket
import time
from threading import Thread

publishing_clients = {}


class message_class:
    len_start_bytes = 8
    start_bytes = b'rosmini!'
    len_header_bytes = 2
    # header bytes should be the len of the topic name + if it is publishing or subscribing (0 or 1)




def server(publishing_clients):
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.bind((socket.gethostname(), 6464))
    s.listen(5)

    while True:
        try:
            clientsocket, address = s.accept()
            print(f"Connection from {address} has been established!")
            t = Thread(target=handleClient, args=(clientsocket, address, publishing_clients))
            t.daemon = True
            t.start()
        except KeyboardInterrupt:
            print('Server closed')
            s.close()
            exit()
        except:
            print('Server closed')
            s.close()
            exit()

def handleClient(clientsocket, address, publishing_clients):
    last_send = time.time()
    try:
        data = clientsocket.recv(message_class.len_start_bytes)
        if data:
            if data.decode() == message_class.start_bytes.decode():
                data = clientsocket.recv(message_class.len_header_bytes)
                if data:
                    header = data.decode()
                    topic_name = header[:len(header) - 1]
                    if header[-1] == '0':
                        # publishing
                        if topic_name in publishing_clients:
                            clientsocket.sendall(bytes(publishing_clients[topic_name], 'utf-8'))
                        else:
                            publishing_clients[topic_name] = address
                    else:
                        # subscribing
                        pass
        else:
            clientsocket.close()
            exit()

    except KeyboardInterrupt:
        clientsocket.close()
        exit()
    except:
        clientsocket.close()
        exit()



if __name__ == "__main__":
    
    server(publishing_clients)

rosmini/rosminipy.py
import socket
import time

import ast
import select
from threading import Thread


ANY = "0.0.0.0"
SENDERPORT = 32000
MCAST_ADDR = "237.252.249.227"


def communicate_rosminicore(sock, topic):
    topic_bytes = bytes(topic)
    # size of len_topic_bytes should be 4 bytes (int)
    # zero padding is added to the beginning of the len_topic_bytes
    len_topic_bytes = bytes(str(len(topic)))
    len_topic_bytes = len_topic_bytes.ljust(4, b' ')
    print(len(topic), len_topic_bytes)
    message = b'rosmini!' + b'\n' + len_topic_bytes + b'\n' + topic_bytes
    sock.sendall(b'rosmini!')
    sock.sendall(len_topic_bytes)
    sock.sendall(topic_bytes)
    data, rosminicore_addr = sock.recvfrom(4)
    if data:
        len_topic = int(data.decode())
        data, rosminicore_addr = sock.recvfrom(len_topic, socket.MSG_WAITALL)
        if data:
            ip_port = data.decode()
            pub_ip, pub_port = ast.literal_eval(ip_port)
            print('Topic registered: ' + topic + ' on ' + pub_ip + ':' + str(pub_port))
            return pub_port
    return False

def assign_node(topic):
    sock_rosminicore = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    # sock_rosminicore.bind((socket.gethostname(), 6464))
    sock_rosminicore.connect((socket.gethostname(), 6464))
    pub_port = communicate_rosminicore(sock_rosminicore, topic)
    sock_rosminicore.close()
    if(not pub_port):
        print("Topic could not be registered")
        return
    return pub_port


class Publisher:

    def __init__(self, topic):

        pub_port = assign_node(topic)

        self.MCAST_PORT = pub_port

        self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM,
            socket.IPPROTO_UDP)
        self.sock.bind((ANY, 0))
        self.sock.setsockopt(socket.IPPROTO_IP, socket.IP_MULTICAST_TTL, 255)
        self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)



    def send(self, data):
        len_data = bytes(len(data)).ljust(16, b' ')
        self.sock.sendto(len_data, (MCAST_ADDR, self.MCAST_PORT))
        self.sock.sendto(bytes(data), (MCAST_ADDR, self.MCAST_PORT))

class Subscriber:

    def __init__(self, topic, callback=None):

        pub_port = assign_node(topic)

        self.MCAST_PORT = pub_port

        #create a UDP socket
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
        #allow multiple sockets to use the same PORT number
        self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
        #Bind to the port that we know will receive multicast data
        self.sock.bind((ANY, self.MCAST_PORT))
        #tell the kernel that we are a multicast socket
        self.sock.setsockopt(socket.IPPROTO_IP, socket.IP_MULTICAST_TTL, 255)
        #Tell the kernel that we want to add ourselves to a multicast group
        #The address for the multicast group is the third param
        self.sock.setsockopt(socket.IPPROTO_IP, socket.IP_ADD_MEMBERSHIP,
            socket.inet_aton(MCAST_ADDR) + socket.inet_aton(ANY))

        if(callback):
            self.callback = callback
            t = Thread(target=self.receive_loop)
            t.daemon = True
            t.start()
            

    def receive(self):
        data, addr = self.sock.recvfrom(16)
        len_data = int(data.decode())
        data, addr = self.sock.recvfrom(len_data)
        return data.decode()
    
    def empty(self):
        input = [self.sock]
        while 1:
            inputready, o, e = select.select(input,[],[], 0.0)
            if len(inputready)==0: break
            for s in inputready: s.recv(1)
        
    def receive_loop(self):
        while True:
            data = self.receive()
            self.callback(data)



if __name__ == '__main__':
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument('topic', type=str)
    parser.add_argument('--publish', action='store_true')
    args = parser.parse_args()
    topic = args.topic
    if args.publish:
        pub = Publisher(topic)
        print('sending messages to topic: ' + topic)
        while True:
            print('Sending message')
            pub.send(b'Hi' + b' ' + bytes(str(time.time())))
            time.sleep(0.1)

    
    else:
        def callback(data):
            print(data)
        sub = Subscriber(topic, callback)
        print('getting messages from topic: ' + topic)
        while True:
            pass
    